frontend/public/components/utils/name-value-editor.jsx

```diff
@@ -110,13 +111,105 @@ NameValueEditor.propTypes = {
   ).isRequired,
   updateParentData: PropTypes.func.isRequired,
   configMaps: PropTypes.object,
-  secrets: PropTypes.object
+  secrets: PropTypes.object,
+  envFrom: PropTypes.bool
 };
 NameValueEditor.defaultProps = {
   nameString: 'Key',
   valueString: 'Value',
   addString: 'Add More',
   allowSorting: false,
+  readOnly: false,
+  nameValueId: 0,
+  envFrom: false
+};
+
+NameValueEditor.displayName = 'Name Value Editor';
+
+export const EnvFromEditor = withDragDropContext(class EnvFromEditor extends React.Component {
+  constructor (props) {
+    super(props);
+    this._append = this._append.bind(this);
+    this._change = this._change.bind(this);
+    this._move = this._move.bind(this);
+    this._remove = this._remove.bind(this);
+  }
+
+  _append() {
+    const {updateParentData, nameValuePairs, nameValueId} = this.props;
+    const configMapSecretRef = {name: '', key: ''};
+    updateParentData({nameValuePairs: nameValuePairs.concat([['', {configMapSecretRef}, nameValuePairs.length]])}, nameValueId, EnvType.ENV_FROM);
+  }
+
+  _remove(i) {
+    const {updateParentData, nameValueId} = this.props;
+    const nameValuePairs = _.cloneDeep(this.props.nameValuePairs);
+    nameValuePairs.splice(i, 1);
+    const configMapSecretRef = {name: '', key: ''};
+
+    updateParentData({nameValuePairs: nameValuePairs.length ? nameValuePairs : [['', {configMapSecretRef}]]}, nameValueId, EnvType.ENV_FROM);
+  }
+
+  _change(e, i, type) {
+    const {updateParentData, nameValueId} = this.props;
+    const nameValuePairs = _.cloneDeep(this.props.nameValuePairs);
+    nameValuePairs[i][type === EnvFromPair.Prefix ? EnvFromPair.Prefix : EnvFromPair.Resource] = e.target.value;
+    updateParentData({nameValuePairs}, nameValueId, EnvType.ENV_FROM);
+  }
+
+  _move(dragIndex, hoverIndex) {
+    const {updateParentData, nameValueId} = this.props;
+    const nameValuePairs = _.cloneDeep(this.props.nameValuePairs);
+    const movedPair = nameValuePairs[dragIndex];
+
+    nameValuePairs[dragIndex] = nameValuePairs[hoverIndex];
+    nameValuePairs[hoverIndex] = movedPair;
+    updateParentData({nameValuePairs}, nameValueId, EnvType.ENV_FROM);
+  }
+
+  render () {
+    const {nameValuePairs, readOnly, nameValueId, configMaps, secrets} = this.props;
+    const pairElems = nameValuePairs.map((pair, i) => {
+      const key = _.get(pair, [EnvFromPair.Index], i);
```

Comment:
Is `EnvFromPair.Index` ever not set? I'm a little worried about duplicate keys with the defaulting here.

Reply:
So...this is either always set (in the case of env vars) or never set (in the case on annotations), but there will never be a mixed set of causing dup keys.

Reply:
OK makes sense. Thanks!